### PR TITLE
Remove exceptions from stat_sinks

### PR DESCRIPTION
**Pull request by mathetake**
Commit Message: Remove exceptions from stat_sinks
Additional Description:

This removes the exclusion of stat_sinks extensions from the
allow-list of exceptions.

Risk Level: low
Testing: existing ones
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: source/extensions/stat_sinks/common/statsd/statsd.h [hunk 1]
```diff
@@ -107,7 +107,16 @@ class TcpStatsdSink : public Stats::Sink {
...
+                Stats::Scope& scope, absl::Status& creation_status,
+                const std::string& prefix = getDefaultPrefix());
+
+  /**
+   * This is a wrapper around the constructor to allows it to return a StatusOr.
+   */
+  static absl::StatusOr<std::unique_ptr<TcpStatsdSink>>
+  create(const LocalInfo::LocalInfo& local_info, const std::string& cluster_name,
+         ThreadLocal::SlotAllocator& tls, Upstream::ClusterManager& cluster_manager,
+         Stats::Scope& scope, const std::string& prefix = getDefaultPrefix());
```
Comment (mathetake) [inline source/extensions/stat_sinks/common/statsd/statsd.h:119]: this is following the style of https://github.com/envoyproxy/envoy/pull/37464

File: source/extensions/stat_sinks/common/statsd/statsd.h [hunk 2]
```diff
@@ -107,7 +107,16 @@ class TcpStatsdSink : public Stats::Sink {
...
+                Stats::Scope& scope, absl::Status& creation_status,
+                const std::string& prefix = getDefaultPrefix());
+
+  /**
+   * This is a wrapper around the constructor to return StatusOr.
+   */
+  static absl::StatusOr<std::unique_ptr<TcpStatsdSink>>
+  create(const LocalInfo::LocalInfo& local_info, const std::string& cluster_name,
+         ThreadLocal::SlotAllocator& tls, Upstream::ClusterManager& cluster_manager,
+         Stats::Scope& scope, const std::string& prefix = getDefaultPrefix());
```
Comment (mathetake) [inline source/extensions/stat_sinks/common/statsd/statsd.h:119]: this is following the style of https://github.com/envoyproxy/envoy/pull/37464

**Review by jmarantz — COMMENTED** [review]
looks great with one question.

File: source/extensions/stat_sinks/common/statsd/statsd.h (hunk 2 above)
Comment (mathetake) [reply source/extensions/stat_sinks/common/statsd/statsd.h:119]: i tried but turns out it's not feasible unfortunately - the constructor is accessed inside std::make_unique which blocks it from being private. One workaround for this is to use `std::unique_ptr<T>(new T)` like below but not sure if we want to use `new` just to make this private. wdyt?

```diff
diff --git a/source/extensions/stat_sinks/common/statsd/statsd.cc b/source/extensions/stat_sinks/common/statsd/statsd.cc
index 044d559c2d..c5423ef5be 100644
--- a/source/extensions/stat_sinks/common/statsd/statsd.cc
+++ b/source/extensions/stat_sinks/common/statsd/statsd.cc
@@ -210,8 +210,8 @@ TcpStatsdSink::create(const LocalInfo::LocalInfo& local_info, const std::string&
                       ThreadLocal::SlotAllocator& tls, Upstream::ClusterManager& cluster_manager,
                       Stats::Scope& scope, const std::string& prefix) {
   absl::Status creation_status;
-  auto sink = std::make_unique<TcpStatsdSink>(local_info, cluster_name, tls, cluster_manager, scope,
-                                              creation_status, prefix);
+  auto sink = std::unique_ptr<TcpStatsdSink>(new TcpStatsdSink(
+      local_info, cluster_name, tls, cluster_manager, scope, creation_status, prefix));
   RETURN_IF_ERROR(creation_status);
   return sink;
 }
```

Comment (jmarantz) [reply source/extensions/stat_sinks/common/statsd/statsd.h:119]: It looks like we use `new` in at least one other similar situation, in source/common/upstream/transport_socket_match_impl.cc:18.

Comment (mathetake) [reply source/extensions/stat_sinks/common/statsd/statsd.h:119]: ok let me use new then - than you for taking a look at other places

Comment (mathetake) [reply source/extensions/stat_sinks/common/statsd/statsd.h:114]: https://github.com/envoyproxy/envoy/pull/37491/commits/3c3e1e7ffe59154a75cb90bd7c8577cffd324a9c done

Comment (jmarantz) [reply source/extensions/stat_sinks/common/statsd/statsd.h:114]: thanks!

**Review by jmarantz — APPROVED** [review]

File: source/extensions/stat_sinks/metrics_service/config.cc [hunk 3]
```diff
@@ -25,12 +25,12 @@ MetricsServiceSinkFactory::createStatsSink(const Protobuf::Message& config,
       MessageUtil::downcastAndValidate<const envoy::config::metrics::v3::MetricsServiceConfig&>(
           config, server.messageValidationContext().staticValidationVisitor());
   const auto& grpc_service = sink_config.grpc_service();
-  THROW_IF_NOT_OK(Config::Utility::checkTransportVersion(sink_config));
+  RETURN_IF_NOT_OK_REF(Config::Utility::checkTransportVersion(sink_config));
```
[the remark this answers is not in the document]
Comment (mathetake) [reply source/extensions/stat_sinks/metrics_service/config.cc]: done: https://github.com/envoyproxy/envoy/pull/37491/commits/94bace28044f03dc6ba2ece9a5c3a548f24bfa01

File: test/extensions/stats_sinks/common/statsd/statsd_test.cc [hunk 4]
```diff
@@ -37,9 +37,11 @@ class TcpStatsdSinkTest : public Event::TestUsingSimulatedTime, public testing::
   TcpStatsdSinkTest() {
     cluster_manager_.initializeClusters({"fake_cluster"}, {});
     cluster_manager_.initializeThreadLocalClusters({"fake_cluster"});
-    sink_ = std::make_unique<TcpStatsdSink>(
-        local_info_, "fake_cluster", tls_, cluster_manager_,
-        *(cluster_manager_.active_clusters_["fake_cluster"]->info_->stats_store_.rootScope()));
+    sink_ =
```
[the remark this answers is not in the document]
Comment (mathetake) [reply test/extensions/stats_sinks/common/statsd/statsd_test.cc:40]: yeah I noticed that after it gets approved, and the code was written before you got to know your preference :D will use * from the future PRs - anyways fixed in https://github.com/envoyproxy/envoy/pull/37491/commits/94bace28044f03dc6ba2ece9a5c3a548f24bfa01

**Review by jmarantz — APPROVED** [review]

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**